File: packages_mjs/fetch-auth-config/src/compute-registry.ts
import { ComputeFunctionNotFoundError, ComputeFunctionError } from './errors.js';

export type StartupComputeFn = () => string | Promise<string>;
export type RequestComputeFn = (request: any) => string | Promise<string>;

export class ComputeRegistry {
    private static startupResolvers: Map<string, StartupComputeFn> = new Map();
    private static requestResolvers: Map<string, RequestComputeFn> = new Map();
    private static cache: Map<string, string> = new Map();

    static registerStartup(providerName: string, fn: StartupComputeFn) {
        this.startupResolvers.set(providerName, fn);
    }

    static registerRequest(providerName: string, fn: RequestComputeFn) {
        this.requestResolvers.set(providerName, fn);
    }

    static async resolveStartup(providerName: string): Promise<string> {
        if (this.cache.has(providerName)) {
            return this.cache.get(providerName)!;
        }

        const fn = this.startupResolvers.get(providerName);
        if (!fn) {
            throw new ComputeFunctionNotFoundError(providerName, 'startup');
        }

        try {
            const result = await fn();
            this.cache.set(providerName, result);
            return result;
        } catch (error: any) {
            if (error instanceof ComputeFunctionNotFoundError) throw error;
            throw new ComputeFunctionError(providerName, error);
        }
    }

    static async resolveRequest(providerName: string, request: any): Promise<string> {
        const fn = this.requestResolvers.get(providerName);
        if (!fn) {
            throw new ComputeFunctionNotFoundError(providerName, 'request');
        }

        try {
            return await fn(request);
        } catch (error: any) {
            if (error instanceof ComputeFunctionNotFoundError) throw error;
            throw new ComputeFunctionError(providerName, error);
        }
    }
}

File: packages_py/fetch_auth_config/src/fetch_auth_config/compute_registry.py
import asyncio
from typing import Dict, Callable, Any, Optional, Union, Awaitable
from .errors import ComputeFunctionNotFoundError, ComputeFunctionError

# Types
StartupComputeFn = Callable[[Any], Union[str, Awaitable[str]]]
RequestComputeFn = Callable[[Any], Union[str, Awaitable[str]]]

class ComputeRegistry:
    _startup_resolvers: Dict[str, StartupComputeFn] = {}
    _request_resolvers: Dict[str, RequestComputeFn] = {}

    @classmethod
    def register_startup(cls, provider_name: str):
        def decorator(fn: StartupComputeFn):
            cls._startup_resolvers[provider_name] = fn
            return fn
        return decorator

    @classmethod
    def register_request(cls, provider_name: str):
        def decorator(fn: RequestComputeFn):
            cls._request_resolvers[provider_name] = fn
            return fn
        return decorator

    @classmethod
    async def resolve_startup(cls, provider_name: str, app: Any) -> str:
        fn = cls._startup_resolvers.get(provider_name)
        if not fn:
            raise ComputeFunctionNotFoundError(provider_name, "startup")
        
        try:
            result = fn(app)
            if asyncio.iscoroutine(result):
                result = await result
            return str(result)
        except Exception as e:
            if isinstance(e, ComputeFunctionNotFoundError):
                raise e
            raise ComputeFunctionError(provider_name, e)

    @classmethod
    async def resolve_request(cls, provider_name: str, request: Any) -> str:
        fn = cls._request_resolvers.get(provider_name)
        if not fn:
            raise ComputeFunctionNotFoundError(provider_name, "request")
        
        try:
            result = fn(request)
            if asyncio.iscoroutine(result):
                result = await result
            return str(result)
        except Exception as e:
             if isinstance(e, ComputeFunctionNotFoundError):
                raise e
             raise ComputeFunctionError(provider_name, e)

# Module-level aliases for cleaner imports
register_startup = ComputeRegistry.register_startup
register_request = ComputeRegistry.register_request

